Mathlib/Geometry/Manifold/LocalDiffeomorph.lean

```diff
@@ -0,0 +1,246 @@
+* conversely, if `f` is `C^n` at `x` and `mfderiv I J n f x` is a linear isomorphism,
+`f` is a local diffeomorphism at `x`.
+* if `f` is a local diffeomorphism, each differential `mfderiv I J n f x`
+is a continuous linear isomorphism.
+* Conversely, if `f` is `C^n` and each differential is a linear isomorphism,
+`f` is a local diffeomorphism.
+
+## Implementation notes
+
+This notion of diffeomorphism is needed although there is already a notion of local structomorphism
+because structomorphisms do not allow the model spaces `H` and `H'` of the two manifolds to be
+different, i.e. for a structomorphism one has to impose `H = H'` which is often not the case in
+practice.
+
+## Tags
+local diffeomorphism, manifold
+
+-/
+
+open Manifold Set TopologicalSpace
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  {H : Type*} [TopologicalSpace H]
+  {G : Type*} [TopologicalSpace G]
+  (I : ModelWithCorners 𝕜 E H) (J : ModelWithCorners 𝕜 F G)
+  (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+  (N : Type*) [TopologicalSpace N] [ChartedSpace G N] (n : ℕ∞)
+
+section PartialDiffeomorph
+/-- A partial diffeomorphism on `s` is a function `f : M → N` such that `f` restricts to a
+diffeomorphism `s → t` between open subsets of `M` and `N`, respectively.
+This is an auxiliary definition and should not be used outside of this file. -/
+structure PartialDiffeomorph extends PartialEquiv M N where
+  open_source : IsOpen source
+  open_target : IsOpen target
+  contMDiffOn_toFun : ContMDiffOn I J n toFun source
+  contMDiffOn_invFun : ContMDiffOn J I n invFun target
+
+/-- Coercion of a `PartialDiffeomorph` to function.
+Note that a `PartialDiffeomorph` is not `FunLike` (like `PartialHomeomorph`),
+as `toFun` doesn't determine `invFun` outside of `target`. -/
+instance : CoeFun (PartialDiffeomorph I J M N n) fun _ => M → N :=
+  ⟨fun Φ => Φ.toFun⟩
+
+-- Add the very basic API we need.
+namespace PartialDiffeomorph
+variable (Φ : PartialDiffeomorph I J M N n) (hn : 1 ≤ n)
+
+/-- A partial diffeomorphism is also a local homeomorphism. -/
+def toPartialHomeomorph : PartialHomeomorph M N :=
+  {
+    toPartialEquiv := Φ.toPartialEquiv
+    open_source := Φ.open_source
+    open_target := Φ.open_target
+    continuousOn_toFun := Φ.contMDiffOn_toFun.continuousOn
+    continuousOn_invFun := Φ.contMDiffOn_invFun.continuousOn
+  }
+
+/-- The inverse of a local diffeomorphism. -/
+protected def symm : PartialDiffeomorph J I N M n :=
+  {
+    toPartialEquiv := Φ.toPartialEquiv.symm
+    open_source := Φ.open_target
+    open_target := Φ.open_source
+    contMDiffOn_toFun := Φ.contMDiffOn_invFun
+    contMDiffOn_invFun := Φ.contMDiffOn_toFun
+  }
+
+protected theorem contMDiffOn : ContMDiffOn I J n Φ Φ.source :=
+  Φ.contMDiffOn_toFun
+
+protected theorem contMDiffOn_symm : ContMDiffOn J I n Φ.symm Φ.target :=
+  Φ.contMDiffOn_invFun
```

Comment:
You shouldn't need this one, as well as the other `symm` versions below: just call `Φ.symm.contMDiffOn`, which isn't longer.

Reply:
Fair enough. I went by comparing with PartialHomeomorph, which has almost all lemmas also for `symm`...